### DISCUSSION TITLE
Self-hosted GFS results different from Open-Meteo API.

### DISCUSSION THREAD

**Discussion by wirrell** [open-meteo/open-meteo#221] [Q&A] [answered]
I'm self-hosting a copy of the open-meteo API on an Azure VM using the prebuilt Ubuntu Jammy package. I am running the latest version available through apt (0.1.211).

I am interested in hosting my own ERA-5 historical and GFS forecast mirrors, and have done the following steps to set up my copy.

1. Downloaded ERA-5 year historicals for 2000-2022 using `openmeteo-api  download-era5 era5 --year YEAR -k KEY`
2. Downloaded and processed the DEM files following the DEM section in [getting-started.md](https://github.com/open-meteo/open-meteo/blob/main/docs/getting-started.md)
3. Setup a cronjob with the following commands:
`0 */8 * * *  /usr/local/bin/openmeteo-api download-era5 era5 --cdskey 176509:7c1ad77b-922d-47eb-8df5-7d386268f438 > ~/log/era5.log 2>&1 || cat ~/log/era5.log`
`40 3,9,15,21 * * * /usr/local/bin/openmeteo-api download-gfs gfs025 --skip-existing --only-variables temperature_2m,precipitation,shortwave_radiation`
`40 3,9,15,21 * * * /usr/local/bin/openmeteo-api download-gfs gfs013 --skip-existing --only-variables temperature_2m,precipitation,shortwave_radiation`
Note: I understand that temp and shortwave radiation are gfs013 variables only and fall through on gfs025, and vice versa.

My issue is that when I submit a query for the same lat lon on my VM and the open-meteo server, I get different results.

As an example, my VM returns, for the query
`http://my.api.ip/v1/gfs?daily=precipitation_sum&timezone=auto&latitude=39.0&longitude=-89.0&forecast_days=16&past_days=0&cell_selection=nearest&elevation=nan` I get
{'latitude': 39.001022, 'longitude': -88.98959, 'generationtime_ms': 0.8820295333862305, 'utc_offset_seconds': -21600, 'timezone': 'America/Chicago', 'timezone_abbreviation': 'CST', 'daily_units': {'time': 'iso8601', 'precipitation_sum': 'mm'}, 'daily': {'time': ['2023-02-15', '2023-02-16', '2023-02-17', '2023-02-18', '2023-02-19', '2023-02-20', '2023-02-21', '2023-02-22', '2023-02-23', '2023-02-24', '2023-02-25', '2023-02-26', '2023-02-27', '2023-02-28', '2023-03-01', '2023-03-02'], '**precipitation_sum': [0.0, 8.0, 0.3, 18.4, 0.0, 0.0, 0.0, 0.4, 8.2, 4.3, 0.0, 0.3, 0.0, 8.4, 1.4, 6.1]**}}

Whereas for the open-meteo server for query:

`https://api.open-meteo.com/v1/gfs?daily=precipitation_sum&timezone=auto&latitude=39.0&longitude=-89.0&forecast_days=16&past_days=0&cell_selection=nearest&elevation=nan`
{'latitude': 39.001022, 'longitude': -88.98959, 'generationtime_ms': 0.5719661712646484, 'utc_offset_seconds': -21600, 'timezone': 'America/Chicago', 'timezone_abbreviation': 'CST'**, 'elevation': 168.0,** 'daily_units': {'time': 'iso8601', 'precipitation_sum': 'mm'}, 'daily': {'time': ['2023-02-15', '2023-02-16', '2023-02-17', '2023-02-18', '2023-02-19', '2023-02-20', '2023-02-21', '2023-02-22', '2023-02-23', '2023-02-24', '2023-02-25', '2023-02-26', '2023-02-27', '2023-02-28', '2023-03-01', '2023-03-02'], '**precipitation_sum': [0.0, 2.2, 0.0, 0.1, 0.0, 0.7, 2.1, 17.8, 0.2, 0.0, 0.0, 0.0, 3.1, 27.5, 0.0, 0.0]**}}

I have put the differences in bold, which include the elevation field.

There are also differences in results for solar radiation and temperature queries. I also have minor differences in ERA5 historical queries for the same location, though in this instance shortwave_radiation values are identical for both and rain values have < 1mm difference.

Given that my VM does not return the elevation field, I am initially guessing that this may be something to do with downscaling on the open-meteo server which I am missing on my VM.  I understand that the open-meteo GFS API uses HRRR, however as I understand it that would only affect the first 48 hours of values. Are there any auxiliary datasets which I am not downloading that might be needed to match results? Or additionally, any further setup steps which I have neglected?

**Comment by wirrell** [comment]
Here's a brief GFS side by of my VM versus the open-meteo API in case a pattern is obvious -


|            |   VM TMIN |  VM  TMAX | VM  SRAD | VM  RAIN | OMAPI  TMIN | OMAPI  TMAX | OMAPI  SRAD | OMAPI  RAIN |
|:-----------|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|
| 2023-02-15 |    3.9 |   13.4 |  24.22 |    0   |    5.2 |   14.9 |  15.03 |    0   |
| 2023-02-16 |    4.8 |   13.2 |   3.77 |    3.8 |   -1.2 |    5.9 |   2.08 |    2.2 |
| 2023-02-17 |    7.9 |   15.7 |  14.9  |    4.5 |   -2.5 |    1.8 |  15.68 |    0   |
| 2023-02-18 |   -2.7 |   14.6 |   8.21 |   18.2 |   -2.3 |    7.8 |  11.66 |    0.1 |
| 2023-02-19 |   -5.1 |    0.7 |  16.11 |    0.2 |    0.6 |   12.1 |  14.94 |    0   |
| 2023-02-20 |   -3.2 |    7.4 |  11.07 |    0   |    5.5 |   10.9 |  10.35 |    0.7 |
| 2023-02-21 |    1.8 |   12.8 |  15.49 |    0   |    4.3 |    8.4 |   2.07 |    2.1 |
| 2023-02-22 |    8.5 |   15   |  14.04 |    0   |    4.9 |   16.7 |   2.23 |   17.8 |
| 2023-02-23 |    4.5 |   10.3 |   1.39 |    4.1 |   -1.3 |   19.6 |  16.31 |    0.2 |
| 2023-02-24 |   -1.4 |    4.3 |   1.45 |    8.8 |   -7.3 |   -1.2 |  17.4  |    0   |
| 2023-02-25 |   -3.3 |    5.3 |  14.81 |    0   |   -5.9 |    1.1 |  12.33 |    0   |
| 2023-02-26 |   -1.8 |   13   |  15.7  |    0   |   -3.6 |    2.7 |   9.51 |    0   |
| 2023-02-27 |   -2.8 |    5.8 |  17.31 |    0.3 |    0.5 |   13.6 |   0.78 |    3.1 |
| 2023-02-28 |   -0.4 |    3   |   0.67 |    7.8 |   -0.1 |   16.2 |   7.11 |   27.5 |
| 2023-03-01 |   -2.7 |    3   |   5.72 |    0.8 |   -1.9 |    5.5 |  17.54 |    0   |
| 2023-03-02 |    2.1 |   14.8 |   9.32 |    7.2 |   -0.4 |   13   |  17.54 |    0   |

**Comment by phillipseamore** [comment, marked as answer]
You can select the model in the forecast API, here's just gfs_global:
`https://api.open-meteo.com/v1/forecast?daily=precipitation_sum&timezone=auto&latitude=39.0&longitude=-89.0&past_days=0&cell_selection=nearest&elevation=nan&models=gfs_global&start_date=2023-02-15&end_date=2023-03-02`

`{"latitude":38.952164,"longitude":-88.94531,"generationtime_ms":1.7690658569335938,"utc_offset_seconds":-21600,"timezone":"America/Chicago","timezone_abbreviation":"CST","elevation":170.0,"daily_units":{"time":"iso8601","precipitation_sum":"mm"},"daily":{"time":["2023-02-15","2023-02-16","2023-02-17","2023-02-18","2023-02-19","2023-02-20","2023-02-21","2023-02-22","2023-02-23","2023-02-24","2023-02-25","2023-02-26","2023-02-27","2023-02-28","2023-03-01","2023-03-02"],"precipitation_sum":[0.00,5.00,0.00,0.10,0.00,0.70,2.10,17.80,0.20,0.00,0.00,0.00,3.10,27.50,0.00,0.00]}}`

Do you have DEM files in `data/omfile-dem90` ? Did you test restarting open-meteo after adding the DEM's (`sudo systemctl restart openmeteo-api`)?

**Reply by wirrell** [reply]
Thanks for your response. I have the DEM files and I had tried restarting the service. I was still experiencing the problem even well calling the gfs_global using the forecast API.

I have now fixed the issue. What fixed it was deleting my omfile-gfs013 and omfile-gfs025 directories and re-downloading. I think the issue may have been caused by a cron-initiated gfs download being interrupted by a server restart during the 'converting' stage. I'm not familiar with the internals but maybe this messed up the indexing? Either way, now sorted by deleting the omfile gfs directories and redownloading.

**Reply by patrick-zippenfenig** [reply]
The conversion process should be safe to interrupt. It writes to temporary files with a `~` at the end (e.g. `temperature_2m_1234.om~`) and uses atomic move operations to overwrite an existing file. You just have to be careful that GFS downloads do not overlap. E.g. the 6z run starts while the 0z run is still running.

**Reply by wirrell** [reply]
@patrick-zippenfenig After playing around with it some more, it seems that when I have the --skip-existing flag in my commands for the cronjob, the download-gfs command is finding the old gfs run files in the download-gfs* dirs, and then using them to assign to the new run timestamps, even if they are from an older run. Removing the --skip-existing flag fixed the issue.

**Reply by patrick-zippenfenig** [reply]
Oh yes. Sorry I did not spot it. I used `--skip-existing` only for debugging the conversion step. Otherwise a new download is triggered for each test run.
